Add template engine ADR (#11900)
JabRef/jabref: docs/decisions/0039-template-engine.md

---
nav_order: 39
parent: Decision Records
---

# Template engine

## Context and Problem Statement

We need to choose a template engine for [custom export filters](https://docs.jabref.org/collaborative-work/export/customexports) and [AI features](https://github.com/JabRef/jabref/pull/11884).

A discussion of template engines was also in [one of the JabRef repos](https://github.com/koppor/jabref/issues/392).

A discussion was raised on StackOverflow ["Velocity vs. FreeMarker vs. Thymeleaf"](https://stackoverflow.com/q/1459426/10037342).

## Decision Drivers

* It should be fast.
* It should be possible to provide templates out of Strings (required by the AI feature).

> * It should have short and understandable syntax. Especially. Work with unset fields and empty Optionals should be easy.

Reviewer: Fix grammar. Second and third sentence can be combined.

---
nav_order: 39
parent: Decision Records
---

# Template engine

## Context and Problem Statement

We need to choose a template engine for [custom export filters](https://docs.jabref.org/collaborative-work/export/customexports) and [AI features](https://github.com/JabRef/jabref/pull/11884).

A discussion of template engines was also in [one of the JabRef repos](https://github.com/koppor/jabref/issues/392).

A discussion was raised on StackOverflow ["Velocity vs. FreeMarker vs. Thymeleaf"](https://stackoverflow.com/q/1459426/10037342).

## Decision Drivers

> * It should support different formats, in case JabRef changes in the future. Also, AI features require just plain text manipulation.

Reviewer: ```suggestion
* Text output has to be supported.
```

I don't think we work on binary output? We currently support HTML and other text output formats.

Maybe you mean

* Special escape modes for special languages (e.g., XML or JavaScript) are nice to be future proof.

(After reading the Thymeleaf link. They have a different syntax for each dialect)

But I don't think, this is important.

> * It should have all necessary constructs from programming languages: loops, conditions, variables. It should support list or object (map) variables.

Reviewer: Why this? I would like to have it cooking coffee for me 😅.  "necessary constructs" is too vague. Enumarate the constructs and say why (!). You can say that this is required by the AI feature. This is OK.

Author: I meant, that we don't need a template engine that just does find-and-replace, but something more powerfull

Author: Though, any template engine must be more powerful, otherwise, it's not a template engine

Reviewer: Maybe, rephrase to

* It has to offer loops to enable iterating on a list of entries.

I am very conservative here. My thought is: Is the decision driver measurable somehow. Moreover, "turing compeleness" can be achieved with very simple constructs. Therefore, one needs to be exact in the wishes.

Author: Maybe I'll remove it

> * It should be embeddable and small. AI features require that templates could be generated out of simple `String` type.

Reviewer: What does that mean? "small"? 1x1cm? Do you mean "fast"? Do you mean fast startup time?

```suggestion
* It should be fast. 
* It should be possible to provide templates out of Strings (required by the AI feature)
```

## Considered Options

* Apache Velocity
* Apache FreeMarker
* Thymeleaf

## Decision Outcome

Chosen option: "{title of option 1}", because
{justification. e.g., only option, which meets k.o. criterion decision driver | which resolves force {force} | … | comes out best (see below)}.

## Pros and Cons of the Options

### Apache Velocity

Main page: <https://velocity.apache.org/>.
User guide: <https://velocity.apache.org/engine/devel/user-guide.html>.

> Developer guide: <https://velocity.apache.org/engine/devel/developer-guide.html>.

Reviewer: Please use Markdown lists, otherwise, it is rendred badly

Use `-` as lsit symbol, then GitHub seperates this list from the other list.

Add following to the end of the file

    <!-- markdownlint-disable-file MD004 -->

---

Do that also for the other options.

Reviewer: Please adress this!

Author: But everywhere else we use `*` for lists

* Good, because supports plain text templating.
* Good, because it is possible to use `String` as a template.

> * Good, because it is minimalistic.

Reviewer: What does that mean? The syntax? The Java code? The size of the `jar` file?

* Bad, because {argument d}

> * Bad, because not in active development (but support is still available).

Reviewer: ```suggestion
* Bad, because it is in maintenance mode.
```
